wgpu-types/src/lib.rs

```diff
@@ -939,7 +941,18 @@ impl Default for FilterMode {
     }
 }
 
-#[repr(C)]
+#[repr(u8)]
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+#[cfg_attr(feature = "trace", derive(Serialize))]
+#[cfg_attr(feature = "replay", derive(Deserialize))]
+pub enum AnisotropyLevel {
```

Comment:
The advantages here are obvious (i.e. being correct by construction), the disadvantages, however, less so.
We had a lot of iteration on similar aspects of the API in gfx-rs during the years... and quite often the original desire to make the types pretty back-fired in the end. The morale of gfx-rs experience is that we need to strive for simplicity.

In this case, for example, writing this down as `anisotropy_clamp: Some(wgpu:AnisotropyLevel::X2)` isn't any cleaner than `anisotropy_clamp: Some(2)`. And the "correct by construction" aspect is not helping much in practice(!) since the users gets the error instantly upon creating a sampler.

In another case, say you are writing an engine and you don't have that value predefined. You are loading from disk/configuration. Now you have 2 choices:
  1. make your configuration fully use `AnisotropyLevel` enum here, which makes it intrusive. Not always applicable, i.e. if you don't control the format
  2. do a manual `match aniso_level { 0 | 1 => None, 2 => Some(wgpu::AnisotropyLevel::X2), .. }`, which is very annoying.

So in both of these cases the strong enum, despite it being elegant and all that, doesn't bring much to the table in practice.

Also we need to think about what the other parameters are going to be. For example, do you see MSAA level also enumerized in the same way? Unlike unisotropy, MSAA levels have different support on platforms. That means, we don't know at compile time if the value is good anyway. We might as well accept it as an integer without fanciness.

Reply:
That's fair enough. The point about the intrusiveness is an interesting angle that I didn't think about as I'm mainly used to working on end products (where I can see the whole picture). I'll switch it back.